# CRHMR Research Day 2022 Poster Figures

R version 4.0.2

UMAP gene expression heatmaps of GBM subtype genes for both human and retrotransposon-mapped datasets (technically only needed to do the human-genome mapped dataset, since the other dataset will have the same information). 

```{r message=FALSE, warning=FALSE, cache=TRUE, cache.comments=FALSE, collapse=TRUE }
#load libraries and scripts
.libPaths(c("/scratch/samkyy/gete-gbm/renv/library/R-4.0/x86_64-pc-linux-gnu",
            "/tmp/RtmpJsRC8Z/renv-system-library",
            .libPaths()))
.libPaths()

resultsPath <- "~/scratch/gete-gbm/results" # nolint
getwd()

library(Seurat)
library(Matrix)
library(ggplot2)
library(tidyverse)
library(msm)
source("~/scratch/gete-gbm/bin/util.R")
# source("~/scratch/gete-gbm/bin/util_go.R")
source("~/scratch/gete-gbm/bin/util_seurat.R")
source("~/scratch/gete-gbm/bin/util_viz.R")
```

Load the human-genome mapped + retrotransposon-mapped dataset. 

```{r message=FALSE, warning=FALSE, cache=TRUE, cache.comments=FALSE, collapse=TRUE }
## load dataset
gbmsc.gte <- readRDS("~/projects/def-ytanaka/common/te-gbm_proj/analysis/gte_gbmscIntUmap-subtypes.rds") 
```

Check if all markers are present in the count matrix. 

```{r message=FALSE, warning=FALSE, cache=TRUE, cache.comments=FALSE, collapse=TRUE }
match(c('ASCL1', 'OLIG2', 'EGFR', 'GFAP', 'TGFB1', 'RELB'), rownames(gbmsc.gte))
```

Generate the feature plots for each marker. 

```{r message=FALSE, warning=FALSE, cache=TRUE, cache.comments=FALSE, collapse=TRUE }
mkdirToday()
# Save UMAP for genes
size = 4
pdf("/home/samkyy/scratch/gete-gbm/results/2022-06-02/r_GTE_GBMsubtypeMarkers.pdf", width = size*2, height = size*3)
FeaturePlot(gbmsc.gte, 
            reduction = "umap", 
            features = c('ASCL1', 'OLIG2', 'EGFR', 'GFAP', "TGFB1","RELB"), 
            order = TRUE, #replaces sort.cell = TRUE
            min.cutoff = 'q10', 
            label = TRUE)
dev.off()
```

---

Repeat for the human-genome-mapped dataset.

```{r message=FALSE, warning=FALSE, cache=TRUE, cache.comments=FALSE, collapse=TRUE }
rm(gbmsc.gte)
## load dataset
gbmsc.ge <- readRDS("~/projects/def-ytanaka/common/te-gbm_proj/analysis/ge_gbmscIntUmap-subtypes.rds") 
```

```{r message=FALSE, warning=FALSE, cache=TRUE, cache.comments=FALSE, collapse=TRUE }
match(c('ASCL1', 'OLIG2', 'EGFR', 'GFAP', 'TGFB1', 'RELB'), rownames(gbmsc.ge))
```

```{r message=FALSE, warning=FALSE, cache=TRUE, cache.comments=FALSE, collapse=TRUE }
# Save UMAP for genes
size = 4
pdf("/home/samkyy/scratch/gete-gbm/results/2022-06-02/r_GE_GBMsubtypeMarkers.pdf", width = size*2, height = size*3)
FeaturePlot(gbmsc.ge, 
            reduction = "umap", 
            features = c('ASCL1', 'OLIG2', 'EGFR', 'GFAP', "TGFB1","RELB"), 
            order = TRUE, #replaces sort.cell = TRUE
            min.cutoff = 'q10', 
            label = TRUE)
dev.off()
```
